### PR TITLE
Lexicon: unique likes, reposts, follows

### LINKED ISSUE #584: Add temporary uniqueness constraints on some bsky records
- Add a recursive delete for records sharing the same subject URIs for likes, reposts, and follows

This is a placeholder solution until the repo API supports "where" conditions in mutations.

### PR DESCRIPTION
**Pull request by devinivy**
This sets-up tracking of backlinks found in likes, reposts, and follows (but can and will be extended to others).  This backlink tracking is used to avoid creation of duplicate like, repost, and follow records through the `createRecord` method.  Will have a more generic solution to this in the near future.

Resolves #584.

### CODE REVIEW

File: packages/pds/src/db/migrations/20230316T153255273Z-backlinks.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    .createTable('backlink')
+    .addColumn('uri', 'varchar', (col) => col.notNull())
+    .addColumn('path', 'varchar', (col) => col.notNull())
+    .addColumn('linkToUri', 'varchar')
+    .addColumn('linkToDid', 'varchar')
+    .addPrimaryKeyConstraint('backlinks_pkey', ['uri', 'path'])
+    .addCheckConstraint(
+      'backlink_link_to_chk',
+      // Exactly one of linkToUri or linkToDid should be set
+      sql`("linkToUri" is null and "linkToDid" is not null) or ("linkToUri" is not null and "linkToDid" is null)`,
```
Comment (dholms) [inline packages/pds/src/db/migrations/20230316T153255273Z-backlinks.ts:14]: 🆒

File: packages/pds/src/db/migrations/20230316T153255273Z-backlinks.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
+import { Kysely, sql } from 'kysely'
+
+export async function up(db: Kysely<Schema>): Promise<void> {
+  await db.schema
+    .createTable('backlink')
+    .addColumn('uri', 'varchar', (col) => col.notNull())
+    .addColumn('path', 'varchar', (col) => col.notNull())
+    .addColumn('linkToUri', 'varchar')
+    .addColumn('linkToDid', 'varchar')
```
Comment (dholms) [inline packages/pds/src/db/migrations/20230316T153255273Z-backlinks.ts:9]: do we want an index on these?

File: packages/pds/src/services/record/index.ts [hunk 3]
```diff
@@ -2,15 +2,21 @@ import { CID } from 'multiformats/cid'
 import { AtUri } from '@atproto/uri'
 import * as common from '@atproto/common'
 import { WriteOpAction } from '@atproto/repo'
+import {
+  atUriFormat,
+  didFormat,
+} from '@atproto/lexicon/src/validators/primitives'
```
Comment (dholms) [inline packages/pds/src/services/record/index.ts:8]: these verify fns will be in their respective packages after this PR:
https://github.com/bluesky-social/atproto/pull/576

no need to change tho

**Review by dholms — APPROVED** [review]
Nice 💪 

Just the question about indexes on the backlinks table

**Comment by devinivy** [comment]
Added those indexes 👍

**Pull request merged**